price_feed_poller.py

```diff
@@ -53,36 +60,42 @@ def handle_requestUpdate(w3, pricefeedcontract, wrbcontract, account_addr, gas,
       )  
     return receipt['status']
 
-def handle_read_data_request(w3, pricefeedcontract, account_addr, gas, gas_price):
+def handle_completeUpdate(
+    w3,
+    pricefeedcontract,
+    account_addr,
+    gas,
+    gas_price,
+    tx_waiting_timeout_secs,
+    tx_polling_latency_secs
+  ):
+
     # We got a Read DR request!
     print(f"Got data complete request")
-
-    # Check that the accout has enough balance
-    balance = w3.eth.getBalance(account_addr)
-    if balance == 0:
+       
+    try:
+      # Check that the account has enough balance
+      balance = w3.eth.getBalance(account_addr)
+      if balance == 0:
         raise Exception("Account does not have any funds")
 
-    print(f"Got {balance} wei")
-    print(F"Gas price: {gas_price}")
-        
-    # Hardcoded gas since it does not estimate well
-    try:
+      print(f"Got {balance} wei")
+      print(F"Gas price: {gas_price}")
+
       read_id = pricefeedcontract.functions.completeUpdate().transact({
         "from": account_addr,
         "gas": gas,
         "gasPrice": gas_price
       })
-    except:
-      print(f"Failed when calling to", account_addr, ".completeUpdate(). Retrying in next iteration.")
-      return False
 
-    try:     
       # Get receipt of the transaction
-      receipt = w3.eth.waitForTransactionReceipt(read_id , 60, 5)
-    except exceptions.TimeExhausted:
-      print(
-        f"Transaction timeout reached and result read not included in the block. Retrying in next iteration."
-      )
+      try:
+        receipt = w3.eth.waitForTransactionReceipt(read_id, tx_waiting_timeout_secs, tx_polling_latency_secs)
+      except exceptions.TimeExhausted:
+        print(f"Transaction for completing update on {pricefeedcontract.address} is taking too long. Retrying in next iteration.")
+
+    except:
+      print(f"Failed when trying to complete update on {pricefeedcontract.address}. Retrying in next iteration.")
```

Comment:
Same here about nesting.

Reply:
Reason for nesting:

`w3.eth.waitForTransactionReceipt` enters an inner loop, in which `eth_getTransactionReceipt` is called every `tx_polling_latency_secs` seconds. If after `tx_waiting_timeout_secs` the transaction is not mined (ie. received result to `eth_getTransactionReceipt` is null), `w3.eth.waitForTransactionReceipt` will throw exception `TimeExhausted`. We'd rather differentiate this case ("...transaction did not failed but it's taken too long..."), from a situation where gateway/endpoint provider misses to respond JSONRPC calls and/or HTTP timeout is reached (situation controlled by outer try/except block).

Reply:
However, there's a penalty when TimeExhausted is thrown: either `requestUpdate` or `completeUpdate` will be processed again. These repetitions won't fail, but more gas is incurred. 

Corolary: `tx_waiting_timeout_secs` should be set to avg. block time * avg. blocks ellapsed since tx added to mempool and tx gets mined.

Reply:
I'm not suggesting to handle all the exceptions the same, but rather to centrally handle them in a single point of the code by using multiple `except` clauses.

Namely, this:
```python
try:
  # Check that the accout has enough balance
  balance = w3.eth.getBalance(account_addr)

  if balance == 0:
      raise Exception("Account does not have any funds")

  print(f"Got {balance} wei")
  print(f"Gas price: {gas_price}")

  reward = wrbcontract.functions.estimateGasCost(gas_price).call()

  dr_id = pricefeedcontract.functions.requestUpdate().transact({
    "from": account_addr,
    "gas": gas,
    "gasPrice": gas_price,
    "value": reward
  })

  # Get receipt of the transaction   
  receipt = w3.eth.waitForTransactionReceipt(dr_id, tx_waiting_timeout_secs, tx_polling_latency_secs)

except exceptions.TimeExhausted:
  print(f"Transaction for requesting update on {pricefeedcontract.address} is taking too long. Retrying in next iteration.")

except Exception as ex:
  print(f"Failed when trying to update request on {pricefeedcontract.address}. Retrying in next iteration.")
  print(f"Exception: {ex}")
  return False
```